src/main/java/org/dataone/speedbagit/SpeedBagIt.java

```diff
@@ -138,19 +137,51 @@ public SpeedBagIt(double version,
      * Adds a stream of data to the bag.
      *
      * @param file:      The stream representing a file or data that will be placed in the bag
-     * @param bagPath:      The path, relative to the bag root where the file belongs
-     * @param checksum: A MessageDigest object that will hold the checksum
+     * @param bagPath:   The path, relative to the bag root where the file belongs
+     * @param checksum:  A MessageDigest object that will hold the checksum
      * @param isTagFile: Boolean set to True when the file is a tag file
      */
-    public void addFile(InputStream file, String bagPath, MessageDigest checksum, boolean isTagFile) {
+    public void addFile(InputStream file, String bagPath, MessageDigest checksum, boolean isTagFile)
+            throws SpeedBagException {
         logger.debug(String.format("Adding %s to the bag", bagPath));
+        // Check to see if there's a path conflict
+        if (this.hasPathCollisions(bagPath, isTagFile)) {
+            throw new SpeedBagException(
+                    String.format("The tag file with path %s conflicts with another file.", bagPath)
+            );
+        }
         SpeedFile newFile = new SpeedFile(new SpeedStream(file, checksum), bagPath, isTagFile);
-        if (isTagFile) {
-            this.tagFiles.add(newFile);
-        } else {
-            this.dataFiles.add(newFile);
+            if (isTagFile) {
+                this.tagFiles.add(newFile);
+            } else {
+                this.dataFiles.add(newFile);
+            }
+        }
+
+        /**
+         * Checks whether two paths collide, based on their file type (tag vs data file).
+         *
+         * @param path:  Path being checked against the previously added files
+         * @param isTagFile: A flag whether the file is a data file or not (otherwise it will be a tag)
+         */
+        private boolean hasPathCollisions(String path, boolean isTagFile) {
+            if (isTagFile) {
+                for (SpeedFile tagFile : this.tagFiles) {
```

Comment:
This loop seems unnecessarily slow -- why use an iterator whiuch searches in O(n) time when you could use a HashSet which searches in O(1) time? This seems true for both this `tagFiles` structure and more importantly, the `dataFiles` structure below.

Reply:
I briefly considered it but went with the existing lists since a hashset would be an additional data structure in memory of size (tag files + data files) and we wouldn't be getting any use out of it other than in this one spot. If we don't mind that-I can do a quick refactor to use the HashSet-just give me a thumbs up! Since this _is_ a critical piece where performance matters, I can see the strong argument for it.

Reply:
I am suggesting we replace `ArrayList dataFiles` with `HashSet dataFiles`, and same with `tagFiles` throughout. So there would still only be one copy of the list, and memory requirements would be similar. The major difference is in how the list is accessed and appended to. One added advantage of HashSet is that, if you hashed using the path as the key, then the HashSet itself would be able to enforce uniqueness of paths, and you wouldn't even need to access the SpeedFile. I haven't looked through the code comprehensively to see what impact this would have, but with a quick perusal it didn't seem like a major issue to refactor.

Your fix takes care of the existing problem, so it would be perfectly reasonable to fix that and then later refactor for performance. It just jumped out to me. And I do think performance is an issue with this list.

Reply:
378f591569b5365524bf483488561951e8bda0cd has this implemented as `HashMap<String, SpeedFile>`. The three areas of interest are 
1. [Where we add to it](https://github.com/DataONEorg/speed-bagit/pull/28/commits/378f591569b5365524bf483488561951e8bda0cd#diff-d096abe3162e8756944f3ff751fdec9bd2e56e192fa6d358a52dae4f012f4cfcR144)
2. [Where we use it for path collision lookup](https://github.com/DataONEorg/speed-bagit/pull/28/commits/378f591569b5365524bf483488561951e8bda0cd#diff-d096abe3162e8756944f3ff751fdec9bd2e56e192fa6d358a52dae4f012f4cfcR158)
3. [Where we iterate over it](https://github.com/DataONEorg/speed-bagit/pull/28/commits/378f591569b5365524bf483488561951e8bda0cd#diff-d096abe3162e8756944f3ff751fdec9bd2e56e192fa6d358a52dae4f012f4cfcR301)

Slightly unrelated change: I simplified [this](https://github.com/DataONEorg/speed-bagit/pull/28/commits/378f591569b5365524bf483488561951e8bda0cd#diff-d096abe3162e8756944f3ff751fdec9bd2e56e192fa6d358a52dae4f012f4cfcL121) constructor by making use of the _other_ constructor. 

[70c63af](https://github.com/DataONEorg/speed-bagit/pull/28/commits/70c63af43b3f40f0e62f6d44b01693208bdaad5b) has a version bump so that it'll be ready to be shipped off to maven.

If there's anything else that you see LMK. All in all I think this was a really good change to get in